extensions/smallrye-fault-tolerance/deployment/src/test/java/io/quarkus/smallrye/faulttolerance/test/FaultToleranceTest.java

```diff
@@ -0,0 +1,87 @@
+package io.quarkus.smallrye.faulttolerance.test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import javax.inject.Inject;
+
+import org.eclipse.microprofile.faulttolerance.exceptions.CircuitBreakerOpenException;
+import org.eclipse.microprofile.faulttolerance.exceptions.TimeoutException;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.smallrye.faulttolerance.test.asynchronous.AsynchronousBean;
+import io.quarkus.smallrye.faulttolerance.test.bukhead.BulkheadBean;
+import io.quarkus.smallrye.faulttolerance.test.circuitbreaker.CircuitBreakerBean;
+import io.quarkus.smallrye.faulttolerance.test.fallback.FallbackBean;
+import io.quarkus.smallrye.faulttolerance.test.retry.RetryBean;
+import io.quarkus.smallrye.faulttolerance.test.timeout.TimeoutBean;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class FaultToleranceTest {
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(FallbackBean.class, BulkheadBean.class, TimeoutBean.class, RetryBean.class,
+                            CircuitBreakerBean.class, AsynchronousBean.class));
+
+    @Inject
+    FallbackBean fallback;
+    @Inject
+    BulkheadBean bulkhead;
+    @Inject
+    TimeoutBean timeout;
+    @Inject
+    RetryBean retry;
+    @Inject
+    CircuitBreakerBean circuitbreaker;
+    @Inject
+    AsynchronousBean asynchronous;
+
+    @Test
+    public void testFallback() {
+        assertEquals(FallbackBean.RecoverFallback.class.getName(), fallback.ping());
+    }
+
+    @Test
+    public void testBulkhead() {
+        ExecutorService executorService = Executors.newFixedThreadPool(10);
```

Comment:
Perhaps `shutdown()` the executor service at the end of the test?

Reply:
Actually the asserts here happen on different threads, so they don't really count (unless JUnit 5 is significantly different from JUnit 4 in this regard, which I don't know). Probably easiest would be something like this (writing from memory, some tweaks might be necessary, but the general idea should be clear):

```java
AtomicInteger success = new AtomicInteger();
ExecutorService executorService = Executors.newFixedThreadPool(10);
for (int i = 0; i < 20; i++) {
    executorService.submit(() -> {
        int bulk = bulkhead.bulkhead();
        if (bulk <= 5) {
            success.incrementAndGet();
        }
    });
}
executorService.shutdown();
executorService.awaitTermination(5, TimeUnit.SECOND);
assertEquals(20, success.get());
```

Reply:
Assertions are done on the same thread (I check the code) so my impl is correct. I assert 20 times that the bulk is less or equals to 5 and it's pretty the same as you code but with an AtomicInteger to count. I prefere mine as it carry the intent better.

I added the closing of the executorService, thanks for pointing this to me :)

Reply:
Well, the `assertTrue` call is in the lambda that you pass to `executorService.submit` -- unless I'm seriously mistaken, that means the assert will happen on the thread from the executor thread pool, not on the main thread on which the test runs.

Reply:
@Ladicek you're right, the assert exception is thrown inside the task and ignored ... 

But Bulkhead also throws an exception when too many parallel queries run so the test should assert that an exception was thrown or that bulkhead was never more than five.

Reply:
Is the new test implementation OK ?